File: app/src/main/java/org/thoughtcrime/securesms/stickers/StickerSearchRepository.java
package org.thoughtcrime.securesms.stickers;

import android.database.Cursor;

import androidx.annotation.NonNull;
import androidx.annotation.WorkerThread;

import org.signal.core.util.concurrent.SignalExecutors;
import org.thoughtcrime.securesms.components.emoji.EmojiUtil;
import org.thoughtcrime.securesms.database.AttachmentTable;
import org.thoughtcrime.securesms.database.SignalDatabase;
import org.thoughtcrime.securesms.database.StickerTable;
import org.thoughtcrime.securesms.database.StickerTable.StickerRecordReader;
import org.thoughtcrime.securesms.database.model.StickerRecord;
import org.thoughtcrime.securesms.emoji.EmojiSource;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.schedulers.Schedulers;

public final class StickerSearchRepository {

  private final StickerTable    stickerDatabase;
  private final AttachmentTable attachmentDatabase;

  public StickerSearchRepository() {
    this.stickerDatabase    = SignalDatabase.stickers();
    this.attachmentDatabase = SignalDatabase.attachments();
  }

  public @NonNull Single<List<StickerRecord>> searchByEmoji(@NonNull String emoji) {
    if (emoji.isEmpty() || emoji.length() > EmojiSource.getLatest().getMaxEmojiLength()) {
      return Single.just(Collections.emptyList());
    }

    return Single.fromCallable(() -> searchByEmojiSync(emoji));
  }

  public void searchByEmoji(@NonNull String emoji, @NonNull Callback<List<StickerRecord>> callback) {
    SignalExecutors.BOUNDED.execute(() -> {
      callback.onResult(searchByEmojiSync(emoji));
    });
  }

  @WorkerThread
  private List<StickerRecord> searchByEmojiSync(@NonNull String emoji) {
    String              searchEmoji = EmojiUtil.getCanonicalRepresentation(emoji);
    List<StickerRecord> out         = new ArrayList<>();
    Set<String>         possible    = EmojiUtil.getAllRepresentations(searchEmoji);

    for (String candidate : possible) {
      try (StickerRecordReader reader = new StickerRecordReader(stickerDatabase.getStickersByEmoji(candidate))) {
        StickerRecord record = null;
        while ((record = reader.getNext()) != null) {
          out.add(record);
        }
      }
    }

    return out;
  }

  public void getStickerFeatureAvailability(@NonNull Callback<Boolean> callback) {
    SignalExecutors.BOUNDED.execute(() -> {
      callback.onResult(getStickerFeatureAvailabilitySync());
    });
  }

  @WorkerThread
  private Boolean getStickerFeatureAvailabilitySync() {
    try (Cursor cursor = stickerDatabase.getAllStickerPacks("1")) {
      if (cursor != null && cursor.moveToFirst()) {
        return true;
      } else {
        return attachmentDatabase.hasStickerAttachments();
      }
    }
  }

  public interface Callback<T> {
    void onResult(T result);
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/keyboard/sticker/StickerSearchRepository.kt
package org.thoughtcrime.securesms.keyboard.sticker

import androidx.annotation.WorkerThread
import org.signal.core.util.SqlUtil.buildCaseInsensitiveGlobPattern
import org.thoughtcrime.securesms.components.emoji.EmojiUtil
import org.thoughtcrime.securesms.database.EmojiSearchTable
import org.thoughtcrime.securesms.database.SignalDatabase
import org.thoughtcrime.securesms.database.StickerTable
import org.thoughtcrime.securesms.database.StickerTable.StickerPackRecordReader
import org.thoughtcrime.securesms.database.StickerTable.StickerRecordReader
import org.thoughtcrime.securesms.database.model.StickerPackRecord
import org.thoughtcrime.securesms.database.model.StickerRecord

private const val RECENT_LIMIT = 24
private const val EMOJI_SEARCH_RESULTS_LIMIT = 20

class StickerSearchRepository {

  private val emojiSearchTable: EmojiSearchTable = SignalDatabase.emojiSearch
  private val stickerTable: StickerTable = SignalDatabase.stickers

  @WorkerThread
  fun search(query: String): List<StickerRecord> {
    if (query.isEmpty()) {
      return StickerRecordReader(stickerTable.getRecentlyUsedStickers(RECENT_LIMIT)).readAll()
    }

    val maybeEmojiQuery: List<StickerRecord> = findStickersForEmoji(query)
    val searchResults: List<StickerRecord> =
      // Match by title first, then by emoji.
      StickerRecordReader(stickerTable.getStickerPacksByTitle(query)).readAll()
        .plus(emojiSearchTable.query(query, EMOJI_SEARCH_RESULTS_LIMIT)
          .map(::findStickersForEmoji)
          .flatten())

    return maybeEmojiQuery + searchResults
  }

  @WorkerThread
  private fun findStickersForEmoji(emoji: String): List<StickerRecord> {
    val searchEmoji: String = EmojiUtil.getCanonicalRepresentation(emoji)

    return EmojiUtil.getAllRepresentations(searchEmoji)
      .filterNotNull()
      .map { candidate -> StickerRecordReader(stickerTable.getStickersByEmoji(candidate)).readAll() }
      .flatten()
  }
}

private fun StickerRecordReader.readAll(): List<StickerRecord> {
  val stickers: MutableList<StickerRecord> = mutableListOf()
  use { reader ->
    var record: StickerRecord? = reader.next
    while (record != null) {
      stickers.add(record)
      record = reader.next
    }
  }
  return stickers
}

private fun StickerPackRecordReader.readAll(): List<StickerPackRecord> {
  val stickers: MutableList<StickerPackRecord> = mutableListOf()
  use { reader ->
    var record: StickerPackRecord? = reader.next
    while (record != null) {
      stickers.add(record)
      record = reader.next
    }
  }
  return stickers
}
